File: src/forager/ui/widgets/titlebar.py
"""The top bar: forager menu, back/forward buttons, and the gamepad hint."""
from __future__ import annotations
from PySide6.QtCore import Qt, Signal, QSize
from PySide6.QtGui import QFont
from PySide6.QtWidgets import (
    QWidget, QHBoxLayout, QLabel, QPushButton, QApplication, QToolButton, QMenu,
    QButtonGroup,
)

from forager.ui.fonts import UI_FONT
from forager.ui.theme import C, NAV_TAB_QSS
from forager.ui.icons import load_icon as load_bundled_icon
from forager.ui import style


class TitleBar(QWidget):
    settings_requested = Signal()
    update_proton_requested = Signal()
    run_updates_requested = Signal()
    back_requested = Signal()
    store_tab_requested = Signal()
    library_tab_requested = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setFixedHeight(52)
        self.setStyleSheet(style.surface(1) + f" border-bottom: 1px solid {C.COLOR_3};")
        lay = QHBoxLayout(self)
        lay.setContentsMargins(16, 0, 16, 0)
        lay.setSpacing(10)

        logo = QToolButton()
        logo.setText("forager")
        logo.setFont(QFont(UI_FONT, 16, QFont.Weight.Bold))
        logo.setCursor(Qt.CursorShape.PointingHandCursor)
        logo.setToolTip("forager menu")
        logo.setPopupMode(QToolButton.ToolButtonPopupMode.InstantPopup)
        logo.setStyleSheet(style.toolbutton_qss())
        self._main_menu = QMenu(self)
        self._main_menu.addAction("Settings…", self.settings_requested.emit)
        self._main_menu.addAction("Update Proton", self.update_proton_requested.emit)
        self._main_menu.addSeparator()
        self._main_menu.addAction("Quit", QApplication.instance().quit)
        logo.setMenu(self._main_menu)
        lay.addWidget(logo)

        self._back_btn = self._nav_button("arrow-left")
        self._forward_btn = self._nav_button("arrow-right")
        self._back_btn.setToolTip("Back to Library")
        self._forward_btn.setToolTip("Forward")
        self._back_btn.clicked.connect(self.back_requested)
        self._forward_btn.setEnabled(False)
        lay.addWidget(self._back_btn)
        lay.addWidget(self._forward_btn)

        lay.addStretch(1)

        self._tabs = self._build_tabs()
        lay.addWidget(self._tabs, 0, Qt.AlignmentFlag.AlignCenter)

        lay.addStretch(1)

        self._update_pill = QPushButton()
        self._update_pill.setCursor(Qt.CursorShape.PointingHandCursor)
        self._update_pill.setStyleSheet(style.pill_qss())
        self._update_pill.hide()
        self._update_pill.clicked.connect(self.run_updates_requested)
        lay.addWidget(self._update_pill)

        self._controller_hint = QLabel("")
        style.label(self._controller_hint, C.TEXT_DIM, size=11, padding="4px 8px")
        lay.addWidget(self._controller_hint)

    def _build_tabs(self) -> QWidget:
        bar = QWidget()
        lay = QHBoxLayout(bar)
        lay.setContentsMargins(0, 0, 0, 0)
        lay.setSpacing(0)

        group = QButtonGroup(self)
        group.setExclusive(True)
        self._store_tab = self._tab_button("Store", group, self.store_tab_requested)
        self._library_tab = self._tab_button("Library", group, self.library_tab_requested)
        lay.addWidget(self._store_tab)
        lay.addWidget(self._library_tab)
        self._tabs_group = group
        self._library_tab.setChecked(True)
        return bar

    def _tab_button(self, text: str, group: QButtonGroup, signal) -> QPushButton:
        btn = QPushButton(text)
        btn.setCheckable(True)
        btn.setCursor(Qt.CursorShape.PointingHandCursor)
        btn.setStyleSheet(NAV_TAB_QSS)
        btn.clicked.connect(signal)
        group.addButton(btn)
        return btn

    def set_active_tab(self, name: str) -> None:
        if name == "store":
            self._store_tab.setChecked(True)
        else:
            self._library_tab.setChecked(True)

    def _nav_button(self, icon_name: str) -> QPushButton:
        btn = QPushButton()
        btn.setFixedSize(32, 32)
        btn.setIcon(load_bundled_icon(icon_name, C.TEXT))
        btn.setIconSize(QSize(18, 18))
        btn.setCursor(Qt.CursorShape.PointingHandCursor)
        btn.setStyleSheet(style.icon_button_qss())
        return btn

    def set_back_enabled(self, enabled: bool):
        self._back_btn.setEnabled(enabled)

    def set_controller_hint(self, text: str):
        self._controller_hint.setText(text)

    def set_updates(self, outdated: list[str]) -> None:
        if not outdated:
            self._update_pill.hide()
            return
        n = len(outdated)
        self._update_pill.setText(f"{n} update{'s' if n != 1 else ''} available")
        self._update_pill.setToolTip("Updates ready: " + ", ".join(outdated))
        self._update_pill.show()

File: src/ui/widgets/TitleBar.cpp
#include "ui/widgets/TitleBar.h"

#include <QApplication>
#include <QButtonGroup>
#include <QHBoxLayout>
#include <QLabel>
#include <QMenu>
#include <QPushButton>
#include <QToolButton>

#include "ui/Fonts.h"
#include "ui/Icons.h"
#include "ui/Style.h"
#include "ui/Theme.h"

using namespace theme;
using namespace theme::C;

static QPushButton* navTabButton(const QString& text, QButtonGroup* group) {
    auto* btn = style::button(text, "navtab");
    btn->setCheckable(true);
    group->addButton(btn);
    return btn;
}

TitleBar::TitleBar(QWidget* parent)
    : QWidget(parent)
{
    setFixedHeight(52);
    setStyleSheet(QStringLiteral("background-color: %1; border-bottom: 1px solid %2;")
                      .arg(COLOR_1, COLOR_3));

    auto* lay = new QHBoxLayout(this);
    lay->setContentsMargins(16, 0, 16, 0);
    lay->setSpacing(10);

    _logo = new QToolButton();
    _logo->setText("forager");
    _logo->setFont(QFont(fonts::UI_FONT, 16, QFont::Bold));
    _logo->setCursor(Qt::PointingHandCursor);
    _logo->setToolTip("forager menu");
    _logo->setPopupMode(QToolButton::InstantPopup);
    _logo->setStyleSheet(style::toolbuttonQss());
    _mainMenu = new QMenu(this);
    _mainMenu->addAction("Settings\u2026", this, &TitleBar::settingsRequested);
    _mainMenu->addAction("Update Proton", this, &TitleBar::updateProtonRequested);
    _mainMenu->addSeparator();
    _mainMenu->addAction("Quit", qApp, &QApplication::quit);
    _logo->setMenu(_mainMenu);
    lay->addWidget(_logo);

    _backBtn = navButton("arrow-left");
    _forwardBtn = navButton("arrow-right");
    _backBtn->setToolTip("Back to Library");
    _forwardBtn->setToolTip("Forward");
    connect(_backBtn, &QPushButton::clicked, this, &TitleBar::backRequested);
    _forwardBtn->setEnabled(false);
    lay->addWidget(_backBtn);
    lay->addWidget(_forwardBtn);

    lay->addStretch(1);

    auto* tabBar = new QWidget();
    auto* tabLay = style::hbox(tabBar);
    _tabsGroup = new QButtonGroup(this);
    _tabsGroup->setExclusive(true);
    _storeTab = navTabButton("Store", _tabsGroup);
    _libraryTab = navTabButton("Library", _tabsGroup);
    tabLay->addWidget(_storeTab);
    tabLay->addWidget(_libraryTab);
    connect(_storeTab, &QPushButton::clicked, this, &TitleBar::storeTabRequested);
    connect(_libraryTab, &QPushButton::clicked, this, &TitleBar::libraryTabRequested);
    _libraryTab->setChecked(true);
    lay->addWidget(tabBar, 0, Qt::AlignCenter);

    lay->addStretch(1);

    _updatePill = style::button(QString(), "pill");
    _updatePill->hide();
    connect(_updatePill, &QPushButton::clicked, this, &TitleBar::runUpdatesRequested);
    lay->addWidget(_updatePill);

    _controllerHint = new QLabel("");
    style::label(_controllerHint, TEXT_DIM, 11, -1, "padding: 4px 8px;");
    lay->addWidget(_controllerHint);
}

QPushButton* TitleBar::navButton(const QString& iconName) {
    auto* btn = style::button(QString(), "icon");
    btn->setFixedSize(32, 32);
    btn->setIcon(icons::loadIcon(iconName, TEXT));
    btn->setIconSize(QSize(18, 18));
    return btn;
}

void TitleBar::setActiveTab(const QString& name) {
    if (name == "store")
        _storeTab->setChecked(true);
    else
        _libraryTab->setChecked(true);
}

void TitleBar::setBackEnabled(bool enabled) {
    _backBtn->setEnabled(enabled);
}

void TitleBar::setControllerHint(const QString& text) {
    _controllerHint->setText(text);
}

void TitleBar::setUpdates(int count) {
    if (count <= 0) {
        _updatePill->hide();
        return;
    }
    _updatePill->setText(QStringLiteral("%1 update%2 available")
                             .arg(count)
                             .arg(count != 1 ? QStringLiteral("s") : QString()));
    _updatePill->show();
}
